cexec/main.go

```
@@ -1,111 +1,202 @@
 
-	if blockDevice == "" {
-		log.Fatalf("No Block Device speified with Environment Variable [BLOCK_DEVICE]")
+	if s.blockDevice == "" {
+		return errors.New("no Block Device speified with Environment Variable [BLOCK_DEVICE]")
 	}
 
 	// Create the /mountAction mountpoint (no folders exist previously in scratch container)
-	err := os.Mkdir(mountAction, os.ModeDir)
-	if err != nil {
-		log.Fatalf("Error creating the action Mountpoint [%s]", mountAction)
+	if err := os.Mkdir(mountAction, os.ModeDir); err != nil {
+		return fmt.Errorf("error creating the mount point [%s], error: %w", mountAction, err)
 	}
 
 	// Mount the block device to the /mountAction point
-	err = syscall.Mount(blockDevice, mountAction, filesystemType, 0, "")
-	if err != nil {
-		log.Fatalf("Mounting [%s] -> [%s] error [%v]", blockDevice, mountAction, err)
+	if err := syscall.Mount(s.blockDevice, mountAction, s.filesystemType, 0, ""); err != nil {
+		return fmt.Errorf("error mounting [%s] -> [%s], error: %v", s.blockDevice, mountAction, err)
 	}
-	log.Infof("Mounted [%s] -> [%s]", blockDevice, mountAction)
+	log.Info("mounted device successfully", "source", s.blockDevice, "destination", mountAction)
 
-	if chroot != "" {
-		err = MountSpecialDirs()
-		if err != nil {
-			log.Fatal(err)
+	var exitChroot func() error
+	if s.chroot != "" {
+		if s.updateResolvConf {
+			// fix resolv.conf as it normally doesn't work in chroot
+			// backup the original resolv.conf
+			resolv := filepath.Join(mountAction, "etc", "resolv.conf")
+			if err := backupFile(resolv); err == nil {
+				defer func() {
+					// restore the original resolv.conf
+					if err := restoreFile(resolv); err != nil {
+						log.Error("unable to restore resolv.conf", "error", err)
+					}
+				}()
+				// create an empty resolv.conf in the chroot so that it can be bind mounted
+				if _, err := os.Create(resolv); err != nil {
+					log.Info("error creating resolv.conf, resolv.conf will not work in the chroot", "error", err)
```

I'm interpreting this as a silent failure. What are we optimizing for? I'd be in favor of exiting non-zero if that's an option particularly given this requires an explicit opt-in on l.117.

---

The idea was that a failure to "fix" DNS would cause the Action to not try the command anyway. I will just add a return error.